# Welcome to Lab: Simpson's Paradox 🔬

Before this lab section, you learned about Simpson's Paradox and confounding variables in lecture.  This week, you will find Simpson's Paradox through analysis of a dataset in Python, and get some practice writing conditionals for pandas DataFrames! 

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab Teaching Assistant, your Course Aides, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

In [ ]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Peng"
ca1_name = "Sophia"
ca2_name = "Qiran"
ca3_name = ""

# Say hello to each other!
# - Groups of 3 are ideal :)
# - However, groups of 2 or 4 are fine too!
#
# Question of the Day (QOTD) to Ask Your Group: "What is your favorite music artist (besides Taylor Swift)?"
partner1_name = "Shivani"
partner1_netid = "sr80"
partner1_artist = "drake"

partner2_name = "sofia"
partner2_netid = "sofia5"
partner2_artist = "the neighborhood"

partner3_name = "jenny"
partner3_netid = "jarreci2"
partner3_artist = "bad bunny"

partner4_name = "tori"
partner4_netid = "vrlee2"
partner4_artist = "the lumineers"

<hr style="color: #DD3403;">

# Part 1: The GPA Dataset

Many of you have likely come across one of the GPA visualizations found at https://waf.cs.illinois.edu/discovery/gpa/, either out of 
curiosity or the need to investigate a mysterious GenEd class you signed up for. 

Regardless, these visualizations are built on the **GPA Dataset** of UIUC students across all course subjects. Today, you're going to do some analysis and discover a case of Simpson's Paradox within data taken from UIUC courses!

## Load the GPA Dataset

The most recent version of the "GPA Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/gpa.csv
```

Use Python to load this dataset into a DataFrame called `df`:

In [21]:
import pandas as pd

df = pd.read_csv("https://waf.cs.illinois.edu/discovery/gpa.csv")
df

Year    Term YearTerm Subject  Number                   Course Title  \
0      2023  Spring  2023-sp     AAS     100   Intro Asian American Studies   
1      2023  Spring  2023-sp     AAS     100   Intro Asian American Studies   
2      2023  Spring  2023-sp     AAS     100   Intro Asian American Studies   
3      2023  Spring  2023-sp     AAS     200           U.S. Race and Empire   
4      2023  Spring  2023-sp     AAS     215   US Citizenship Comparatively   
...     ...     ...      ...     ...     ...                            ...   
69064  2010  Summer  2010-su    STAT     410  Statistics and Probability II   
69065  2010  Summer  2010-su    STAT     440    Statistical Data Management   
69066  2010  Summer  2010-su     TAM     212          Introductory Dynamics   
69067  2010  Summer  2010-su     TAM     251   Introductory Solid Mechanics   
69068  2010  Summer  2010-su     TAM     335   Introductory Fluid Mechanics   

      Sched Type  A+   A  A-  ...  C+  C  C-  D+  D  D-  F  W  Students  \
0            DIS   0  11   5  ...   0  0   0   0  0   0  1  0        22   
1            DIS   0  17   2  ...   0  0   0   0  0   0  0  1        23   
2            DIS   0  13   2  ...   0  0   1   0  0   0  1  0        21   
3            LCD   6  15   5  ...   0  0   0   0  1   0  1  0        33   
4            LCD  16  12   2  ...   0  0   0   0  0   0  0  0        33   
...          ...  ..  ..  ..  ...  .. ..  ..  .. ..  .. .. ..       ...   
69064        LEC   5  10   2  ...   0  1   3   0  0   0  2  1        31   
69065        LEC   4  12   8  ...   0  0   0   0  0   0  0  0        28   
69066        LEC   0   1   3  ...   5  1   1   0  2   0  1  0        28   
69067        LCD   1   2   2  ...   3  3   2   0  0   1  1  0        21   
69068        LEC   5   8   9  ...   0  4   0   0  1   0  2  0        39   

             Primary Instructor  
0                 Shin, Jeongsu  
1                 Shin, Jeongsu  
2                Lee, Sabrina Y  
3                Sawada, Emilia  
4                  Kwon, Soo Ah  
...                         ...  
69064        Stepanov, Alexei G  
69065              Unger, David  
69066         Morgan, William T  
69067  Ott-Monsivais, Stephanie  
69068          Keane, Richard D  

[69069 rows x 23 columns]

### 🔬 Test Case Checkpoint 🔬

In [2]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df) == 69069 ), "This is not the GPA dataset you're looking for."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Data Cleaning: An Additional Column

The GPA dataset contains raw GPA data and is not the easiest to work with if we want to analyze intricacies such as the average GPA or passing rate.  Luckily, DataFrames are modifiable, so we can **add more columns** based on what questions we want to answer.

The process of modifying a dataset via deletion (cleaning up empty/unwanted values) or addition (adding new columns) is often called **data cleaning**. This is an important concept in Data Science, because you won't always receive your data in the perfect format for your purposes.

With the GPA dataset, we want to investigate one innocent question posed by a theoretical incoming student: 
- *Is it easier to get an A in STAT or CS courses at UIUC?*

To do this, we'll need to first perform some modifications on our loaded DataFrame, `df`. Specifically, we will need to create:
- An `A_Grades` column, containing the total number of students receiving an A+, A, or A- in every course

### Puzzle 1.1: Creating an `A_Grades` Column

Create the new column `A_Grades` that stores the total number of "A"s given in every course.

- We consider an "A" to be any type of A, in other words "A+", "A", or an "A-" are all included.
- In our `df`, the number of students recieving a particular grade in a course is found in the `df['Grade']` column. For example, `df['A']` contains the number of students who recieved an A.

In [29]:
df['A_Grades'] = df["A+"] + df["A"] + df["A-"]



### 🔬 Test Case Checkpoint 🔬

In [30]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert('A_Grades' in df), "Make sure you've named the A_Grades column properly and added it to the DataFrame."
assert(df['A_Grades'].sum() == 2325768), "Double check the values of your A_Grades column."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


# Working with the GPA Dataset
Now that we have that extra column set up, we can perform basic mathematical analysis on the GPA Dataset to get insight towards our question:
- *Is it easier to get an A in STAT or CS courses at UIUC?*

Should be simple, right? Let's just see which subject, STAT or CS, has a greater percentage of A grades.

### Puzzle 1.2: Subject DataFrames

Select only the rows of the GPA dataset `df` with a `Subject` of `STAT`. Assign these rows to a new DataFrame, `df_STAT`. 
- Make sure your result only contains STAT courses!

In [44]:
df_STAT = df[df.Subject == "STAT"]
df_STAT

Year    Term YearTerm Subject  Number                   Course Title  \
2118   2023  Spring  2023-sp    STAT     100                     Statistics   
2119   2023  Spring  2023-sp    STAT     100                     Statistics   
2120   2023  Spring  2023-sp    STAT     107         Data Science Discovery   
2121   2023  Spring  2023-sp    STAT     200           Statistical Analysis   
2122   2023  Spring  2023-sp    STAT     207       Data Science Exploration   
...     ...     ...      ...     ...     ...                            ...   
69061  2010  Summer  2010-su    STAT     100                     Statistics   
69062  2010  Summer  2010-su    STAT     100                     Statistics   
69063  2010  Summer  2010-su    STAT     400   Statistics and Probability I   
69064  2010  Summer  2010-su    STAT     410  Statistics and Probability II   
69065  2010  Summer  2010-su    STAT     440    Statistical Data Management   

      Sched Type  A+    A  A-  ...   C  C-  D+   D  D-   F  W  Students  \
2118         LEC  51  165  50  ...  24  15   7  14  10  19  3       478   
2119         ONL  82  285  83  ...  51  23  17  19   6  41  3       849   
2120         LEC  44  177  36  ...   7   5   2   0   0   8  0       341   
2121         LEC  59   54  23  ...   3   0   2   0   0   1  0       178   
2122         LEC  27   23   8  ...   6   3   1   0   0   3  0        93   
...          ...  ..  ...  ..  ...  ..  ..  ..  ..  ..  .. ..       ...   
69061        LCD   1   13   3  ...   2   2   3   1   1   1  0        39   
69062        LCD   0   14   3  ...   3   0   0   1   0   3  1        28   
69063        LEC   4   15   7  ...   2   2   0   1   0   3  1        47   
69064        LEC   5   10   2  ...   1   3   0   0   0   2  1        31   
69065        LEC   4   12   8  ...   0   0   0   0   0   0  0        28   

       Primary Instructor  A_Grades  
2118    Flanagan, Karle A       266  
2119    Flanagan, Karle A       450  
2120    Flanagan, Karle A       257  
2121           Yu, Albert       136  
2122       Deeke, Julie M        58  
...                   ...       ...  
69061  Hirtz, Nathaniel R        17  
69062    Dalpiaz, David M        17  
69063      Monrad, Ditlev        26  
69064  Stepanov, Alexei G        17  
69065        Unger, David        24  

[1019 rows x 24 columns]

Select only the rows of the GPA dataset `df` with a `Subject` of `CS`. Assign these rows to a new DataFrame, `df_CS`. 
- Make sure your result only contains CS courses!

In [34]:
df_CS = df[df.Subject == "CS"]

### 🔬 Test Case Checkpoint 🔬

In [35]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_CS to only CS courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 1.3: Comparing Overall Percentages

With our two new DataFrames of STAT and CS courses, use the following code cell to determine the **percentage** of A grades recieved in STAT and CS courses, storing them as variables `stat_a` and `cs_a` respectively. 

Print statements have been provided to show the values you calculate. 

As you work, remember:
- To find the **% of As**, divide the total number of `A_Grades` by the total number of `Students` ($\space \frac{A \space Grades}{Students} \space$).
- The `A_Grades` column you made earlier, and the `Students` column containing the total number of students in each course.
- The syntax `sum(df['column_name'])` can be used to add up the values of all rows in a particular column of a DataFrame.
- Your % should be a decimal between 0 and 1.

In [42]:

stat_a = sum(df_STAT["A_Grades"])/sum(df_STAT["Students"])
print(f'Overall % of As in STAT is: {stat_a}')

cs_a = sum(df_CS["A_Grades"])/sum(df_CS["Students"])
print(f'Overall % of As in CS is: {cs_a}')

Overall % of As in STAT is: 0.609647329889628
Overall % of As in CS is: 0.6119729348794934


### 🔬 Test Case Checkpoint 🔬

In [43]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
import math
assert(math.isclose(stat_a, 0.609647329889628)), "The overall percentage of A grades received in STAT courses does not appear to have been correctly calculated."
assert(math.isclose(cs_a, 0.6119729348794934)), "The overall percentage of A grades received in CS courses does not appear to have been correctly calculated."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Good Comparison?
**Q: What conclusion can you take from the overall percentages found above when asking the original question:**
- *Is it significantly easier to get an A in STAT or CS courses at UIUC, or is so close it would be hard to make general conclusions?*

It would be hard to make a general conclusion since there are different factors that play into the result such as difficulty, class sizes, number of students enrolled in CS and STAT, etc.. 

**Q: Given what you've learned about experimental design, what are some reasons (specific to this dataset) you may not trust this conclusion? If you would trust it, explain why.**

The data set is over 10 years old and classes can change which can contribute to not trusting the conclusion. 

## An Extra Consideration

If you look at the `Year` column of our GPA dataset, you might notice that we have some old data in our dataset - all the way back to **2010**! This means we aren't really answering our question from the perspective of a student now. 

If we want to know if it is easier to **currently** get an A in a STAT or CS course, we should control for the **date** of the data by looking at more **recent years** specifically.

### Puzzle 1.4: More DataFrames

Using the code cells below, define four new DataFrames by selecting from rows of our previously created `df_CS` and `df_STAT`. 

- `df_cs_recent`: all `CS` course data in *recent years* ($\space \geq 2021 \space$)
- `df_stat_recent`: all `STAT` course data in *recent years* ($\space \geq 2021 \space$)
- `df_cs_old`: all other, older `CS` course data ($\space < 2021 \space$)
- `df_stat_old`: all other, older `STAT` course data ($\space < 2021 \space$)

Define *recent years* as **any year after and including 2021**, and older years as any year before 2021. 

Feel free to use conditionals OR the `.isin()` syntax you learned in the last lab.

In [45]:
df_stat_recent = df_STAT[df_STAT.Year >= 2021]

In [46]:
df_cs_recent = df_CS[df_CS.Year >= 2021]

In [47]:
df_stat_old = df_STAT[df_STAT.Year < 2021]

In [48]:
df_cs_old = df_CS[df_CS.Year < 2021]

### 🔬 Test Case Checkpoint 🔬

In [49]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_stat_recent' in vars() ), "Make certain to name the recent STAT courses df_stat_recent."
assert( 'df_cs_recent' in vars() ), "Make certain to name the recent CS courses df_cs_recent."
assert( 'df_stat_old' in vars() ), "Make certain to name the old STAT courses df_stat_old."
assert( 'df_cs_old' in vars() ), "Make certain to name the old CS courses df_cs_old."

assert( len(df_stat_recent[df_stat_recent.Year < 2021] ) == 0 ), "Make sure only years after and including 2021 are in the df_stat_recent DataFrame."
assert( len(df_cs_recent[df_cs_recent.Year < 2021] ) == 0 ), "Make sure only years after and including 2021 are in the df_cs_recent DataFrame."
assert( len(df_stat_old[df_stat_old.Year >= 2021] ) == 0 ), "Make sure only years before 2021 are in the df_stat_old DataFrame."
assert( len(df_cs_old[df_cs_old.Year >= 2021] ) == 0 ), "Make sure only years before 2021 are in the df_cs_old DataFrame."

assert( len(df[ df.index.isin(df_stat_recent.index) & df.index.isin(df_stat_old.index) ]) == 0 ), "Check for duplicate values in your df_stat_recent and df_stat_old DataFrames."
assert( len(df[ df.index.isin(df_cs_recent.index) & df.index.isin(df_cs_old.index) ]) == 0 ), "Check for duplicate values in your df_cs_recent and df_cs_old DataFrames."
assert( len(df_cs_old) + len(df_cs_recent) == len(df_CS) ), "You're excluding some rows from df_cs_old or df_cs_recent. Please double check your conditionals."
assert( len(df_stat_old) + len(df_stat_recent) == len(df_STAT) ), "You're excluding some rows from df_stat_old or df_stat_recent. Please double check your conditionals."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 1.5: New Analysis
Now that we've got all the DataFrames setup with GPA data of the `CS` and `STAT` courses separated by recency (2022 or newer being 'recent'), we can do more in-depth analysis to investigate our question. 

In the following code cells, **calculate the percentages** described by the comment in the cell. Your answer should always be a **decimal between 0 and 1**. 

Just like Puzzle 1.3, remember: 
- To find the % of As, divide the total number of As by the total number of students. 
- The `A_Grades` column you made earlier, and the `Students` column containing the total number of students in each course.
- The syntax `sum(df['column_name'])` can be used to add up the values of all rows in a particular column of a DataFrame.

In [51]:
# Percentage of As received in CS in recent years
cs_recent_a = sum(df_cs_recent["A_Grades"])/sum(df_cs_recent["Students"])
print(f'Percentage of As received in CS in recent years: {cs_recent_a}')

Percentage of As received in CS in recent years: 0.7130059975764269


In [50]:
# Percentage of As received in STAT in recent years
stat_recent_a = sum(df_stat_recent["A_Grades"])/sum(df_stat_recent["Students"])
print(f'Percentage of As received in STAT in recent years: {stat_recent_a}')

Percentage of As received in STAT in recent years: 0.6505842595985506


In [52]:
# percentage of As received in CS in older years
cs_old_a = sum(df_cs_old["A_Grades"])/sum(df_cs_old["Students"])
print(f'Percentage of As received in CS in older years: {cs_old_a}')

Percentage of As received in CS in older years: 0.5714531786360225


In [53]:
# percentage of As received in STAT in older years
stat_old_a = sum(df_stat_old["A_Grades"])/sum(df_stat_old["Students"])
print(f'Percentage of As received in STAT in older years: {stat_old_a}')

Percentage of As received in STAT in older years: 0.5975443876015648


### 🔬 Test Case Checkpoint 🔬

In [54]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(cs_recent_a,  0.7130059975764269)), "The overall percentage of A grades received in CS courses recently does not appear to have been correctly calculated."
assert(math.isclose(stat_recent_a, 0.6505842595985506)), "The overall percentage of A grades received in STAT courses recently does not appear to have been correctly calculated."

assert(math.isclose(cs_old_a, 0.5714531786360225)), "The overall percentage of A grades received in CS courses in older years does not appear to have been correctly calculated."
assert(math.isclose(stat_old_a, 0.5975443876015648)), "The overall percentage of A grades received in STAT courses in older years does not appear to have been correctly calculated."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Observe the Results
**Run the following cell** to format all of your answers as a DataFrame.

Keep in mind that "Older" means data from courses **before 2021**, and "Recent" means courses held **during or after 2021**. 

In [55]:
pd.DataFrame([
  {'Older % of A': cs_old_a, 'Recent % of A': cs_recent_a, 'Overall % of A': cs_a},
  {'Older % of A': stat_old_a, 'Recent % of A': stat_recent_a, 'Overall % of A': stat_a}
], index=['CS', 'STAT'])

Older % of A  Recent % of A  Overall % of A
CS        0.571453       0.713006        0.611973
STAT      0.597544       0.650584        0.609647

Notice that when observing the "Overall % of A" grades received:
- It appears that `STAT` and `CS` are **equally difficult** to get an `A_Grades` (61.197% vs. 60.965%).
- However, if you look at the sub-group of the courses held in years of **2021 and later**, we see that `CS` actually has a higher `A_Grades` rate! 

This is **Simpson's Paradox**: a pattern within a population can appear, disappear, or reverse when you look at subpopulations.

In more formal terms, Simpson's Paradox can cause you to observe a pattern reverse when you look at the overall group statistics versus statistics of groups post-stratification. In this case we are stratifying by time.


### Analysis: Reflecting on New Observations 

Now think about how would you now respond differently to the incoming student's question:
- *Is it easier to get an A in STAT or CS courses at UIUC?*

**Q: Which comparison of percentages do you trust more and why? Are there any other potential confounding variables when answering this question that could be investigated further? Respond with at least three full sentences.**

I would look at recent data percenatges as they would reflect a more accurate answer. I would rather look at recent data since the courses are probably still being offered and probably have smiliar material across 2021-2023. Again, some confoinding variables can be the enrollment in CS and STAT majors, difficulty of classes, courses offered, etc. 

<hr style="color: #DD3403;">

# Part 2: Revisiting the Hello Dataset

Enough about GPA (for now). Two weeks ago, you created a series of questions that made up the **Hello Dataset** and completed the survey by answering all of the questions made by students in DISCOVERY. 

Now, we will load this dataset again and briefly answer a few questions with data about YOU! 

## Load the Hello Dataset

The "Hello Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/hello-sp24.csv
```

Use Python to load this dataset into a DataFrame called `df_hello`:

In [72]:
df_hello = pd.read_csv("https://waf.cs.illinois.edu/discovery/hello-sp24.csv")
df_hello

Name Section Do you like sweet or savory food?  \
0    Prof. Karle     AL1                            Savory   
1     Prof. Wade     AL1                            Savory   
2        Raymond     Y17                            Savory   
3          Keiji     Y13                             Sweet   
4            Ash     Y17                             Sweet   
..           ...     ...                               ...   
606     Jonathan     Y20                             Sweet   
607          Leo     Y19                             Sweet   
608       Afnaan     Y01                            Savory   
609       Nathan     Y17                             Sweet   
610        Ellyn     Y2-                            Savory   

    Do you use Windows or Mac?  \
0                          Mac   
1                      Windows   
2                      Windows   
3                      Windows   
4                      Windows   
..                         ...   
606                        Mac   
607                        Mac   
608                        Mac   
609                    Windows   
610                        Mac   

    Do you like doing your homework when it's released or in the end?  \
0                                         Near the end                  
1                                   When it's released                  
2                                   When it's released                  
3                                   When it's released                  
4                                   When it's released                  
..                                                 ...                  
606                                 When it's released                  
607                                       Near the end                  
608                                       Near the end                  
609                                 When it's released                  
610                                       Near the end                  

    Are you a morning person?  \
0        NOT a Morning Person   
1        NOT a Morning Person   
2              Morning Person   
3              Morning Person   
4        NOT a Morning Person   
..                        ...   
606            Morning Person   
607      NOT a Morning Person   
608      NOT a Morning Person   
609            Morning Person   
610      NOT a Morning Person   

    Are you following the data science duo on Instagram? @datascienceduo  \
0                                                  Yes                     
1                                                  Yes                     
2                                                   No                     
3                                                  Yes                     
4                                                   No                     
..                                                 ...                     
606                                                 No                     
607                                                 No                     
608                                                 No                     
609                                                 No                     
610                                                 No                     

    Dairy or non-dairy? Are you an In-State or Out-of-State student?  \
0                 Dairy                                     In-State   
1                 Dairy                                     In-State   
2                 Dairy                                     In-State   
3                 Dairy                                     In-State   
4                 Dairy                                     In-State   
..                  ...                                          ...   
606           Non-dairy                                 Out-of-State   
607               Dairy                                 Out-of-State   
608               Dairy

### 🔬 Test Case Checkpoint 🔬

In [57]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df_hello) >= 600 and len(df_hello) < 700), "This is not the Hello Dataset you're looking for. Check the URL."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Happiness vs. Battery Life

There is a narrative that "Gen Zs are only happy when they're on their phone" -- let's explore the relationship between your happiness and your cell phone's battery life!

With the Hello Dataset, there's two questions with numeric responses that can help us explore this:
- From 1 to 10, how happy are you right now?
- What is your phone battery percentage currently at?

### Puzzle 2.1: Observation Subsets

In this situation, let's define three levels of happiness:
- You're **very happy** if you answered "how happy are you right now?" with a score of 9 or higher,
- You're **unhappy** if you answered "how happy are you right now?" with a score of less than 3,
- Otherwise, you're just **happy**.

From this, create three DataFrames that contain subsets of the Hello Dataset: 
- `df_very_happy`: including everyone who is "very happy" ($\space \geq 9 \space$)
- `df_happy`: including everyone who is "happy" ($\space <9 $ and $\geq 3 \space$)
- `df_unhappy`: including everyone who is "unhappy" ($\space < 3 \space$)

The `happiness_question` question has been provided to you as a **string** for ease of DataFrame column access. 

In [88]:
happiness_question = 'From 1 to 10, how happy are you right now?'

df_very_happy = df_hello[df_hello[happiness_question] >= 9]
df_very_happy

Name   Section Do you like sweet or savory food?  \
6             Max       Y02                            Savory   
92      Michelle        Y03                             Sweet   
107       Nathan        Y03                            Savory   
115         Emily       Y03                            Savory   
129          Cici       Y12                             Sweet   
130     CA Qiran        Y02                             Sweet   
131        Nathan       Y14                             Sweet   
134         Affan       Y14                            Savory   
145      Samantha       Y17                            Savory   
184       William       Y10                             Sweet   
212          Tony       Y06                            Savory   
221          Xing       Y15                             Sweet   
229     CA Anagha       Y--                            Savory   
251          Adam       Y09                             Sweet   
266         Paige       Y14                             Sweet   
276          Neil       Y10                            Savory   
280         Danny       Y12                             Sweet   
281      Yongkang       Y12                            Savory   
287          Nick       Y24                             Sweet   
299          Zibo       Y05                             Sweet   
315         Simon       Y17                            Savory   
316        Yiming       Y10                             Sweet   
330         David       Y18                            Savory   
331         Shlok       Y07                             Sweet   
332     Sebastian       Y08                             Sweet   
341          Isha       Y01                             Sweet   
355         Ethan       Y20                            Savory   
364         Naomi       Y11                            Savory   
375        Galen        Y09                             Sweet   
387        Shalin       Y15                             Sweet   
417         Ethan       Y18                            Savory   
420        Vishal       Y03                             Sweet   
456        Nathan       Y01                            Savory   
476        Andrew       Y14                            Savory   
480  TA Evangelos  Y07, Y13                            Savory   
484           Gil       Y13                            Savory   
485    Steven Cai       Y18                             Sweet   
496        Zhilin       Y17                             Sweet   
497          Nana       Y03                            Savory   
500         Casey       Y01                            Savory   
510        Yaxuan       Y13                             Sweet   
514    Jiayang Xu       Y06                             Sweet   
529         Jenny       Y11                             Sweet   
530        Justin       Y20                            Savory   
540         Shiyi       Y12                            Savory   
543         Kevin       Y17                            Savory   
546       Anthony       Y15                            Savory   
561         Vedha       Y14                             Sweet   
571        CA Xin       Y11                             Sweet   
583     CA Anchal       Y04                            Savory   
599          ecat       Y20                            Savory   
601        Vedant       Y10                            Savory   
607           Leo       Y19                             Sweet   
608        Afnaan       Y01                            Savory   

    Do you use Windows or Mac?  \
6                      Windows   
92                         Mac   
107                    Windows   
115                        Mac   
129                        Mac   
130                        Mac   
131                    Windows   
134                        Mac   
145                        Mac   
184                        Mac   
212                        Mac   
221                        Ma

In [97]:
df_happy = df_hello[(df_hello[happiness_question] >= 3) & (df_hello[happiness_question] <9)]
df_happy

Name Section Do you like sweet or savory food?  \
0    Prof. Karle     AL1                            Savory   
1     Prof. Wade     AL1                            Savory   
2        Raymond     Y17                            Savory   
3          Keiji     Y13                             Sweet   
4            Ash     Y17                             Sweet   
..           ...     ...                               ...   
602        Tasso     Y12                            Savory   
603      Sathvik     Y20                            Savory   
606     Jonathan     Y20                             Sweet   
609       Nathan     Y17                             Sweet   
610        Ellyn     Y2-                            Savory   

    Do you use Windows or Mac?  \
0                          Mac   
1                      Windows   
2                      Windows   
3                      Windows   
4                      Windows   
..                         ...   
602                    Windows   
603                    Windows   
606                        Mac   
609                    Windows   
610                        Mac   

    Do you like doing your homework when it's released or in the end?  \
0                                         Near the end                  
1                                   When it's released                  
2                                   When it's released                  
3                                   When it's released                  
4                                   When it's released                  
..                                                 ...                  
602                                       Near the end                  
603                                       Near the end                  
606                                 When it's released                  
609                                 When it's released                  
610                                       Near the end                  

    Are you a morning person?  \
0        NOT a Morning Person   
1        NOT a Morning Person   
2              Morning Person   
3              Morning Person   
4        NOT a Morning Person   
..                        ...   
602      NOT a Morning Person   
603      NOT a Morning Person   
606            Morning Person   
609            Morning Person   
610      NOT a Morning Person   

    Are you following the data science duo on Instagram? @datascienceduo  \
0                                                  Yes                     
1                                                  Yes                     
2                                                   No                     
3                                                  Yes                     
4                                                   No                     
..                                                 ...                     
602                                                 No                     
603                                                 No                     
606                                                 No                     
609                                                 No                     
610                                                 No                     

    Dairy or non-dairy? Are you an In-State or Out-of-State student?  \
0                 Dairy                                     In-State   
1                 Dairy                                     In-State   
2                 Dairy                                     In-State   
3                 Dairy                                     In-State   
4                 Dairy                                     In-State   
..                  ...                                          ...   
602               Dairy                                     In-State   
603               Dairy                                     In-State   
606           Non-dairy

In [90]:
df_unhappy = df_hello[df_hello[happiness_question] < 3]
df_unhappy

Name  Section Do you like sweet or savory food?  \
31               CA Ram      Y15                            Savory   
78                Saint      Y13                            Savory   
81                Anika      Y03                            Savory   
86             CA Amaan      Y12                            Savory   
94                Chris      Y19                            Savory   
111               Blake      Y19                            Savory   
126             Charley      Y08                             Sweet   
128              Tracie      Y11                            Savory   
169              Aditya      Y08                             Sweet   
179                Zach      Y16                             Sweet   
207                Chen      Y12                             Sweet   
228               Sonia      Y03                             Sweet   
243                  AJ      Y10                            Savory   
264               Kaiti      Y09                            Savory   
293               Haoyu      Y16                             Sweet   
303              Alyssa      Y04                            Savory   
305              Blythe      Y20                             Sweet   
351              Steven      Y12                            Savory   
359               Devon      Y11                             Sweet   
399               Shaun      Y16                             Sweet   
424             Cynthia      Y06                             Sweet   
439              Andrew      Y18                            Savory   
441               Jerry      Y08                             Sweet   
462               Isaac      Y12                             Sweet   
466        TA Chenyang   Y05 Y15                            Savory   
483              Soumil      Y18                             Sweet   
492  Fiyinfoluwa (fifi)      Y18                            Savory   
503              Izhaan      Y17                            Savory   
508             Leonard      Y08                             Sweet   
509                judy      Y13                             Sweet   
524     Marc Quillin Jr      Y02                            Savory   
535                 Ben      Y05                            Savory   
537               Yifei      Y02                             Sweet   
553                Marc      Y15                            Savory   
567             Neal Wu      Y19                             Sweet   
605            LaKeisha      Y15                            Savory   

    Do you use Windows or Mac?  \
31                         Mac   
78                     Windows   
81                         Mac   
86                         Mac   
94                         Mac   
111                    Windows   
126                    Windows   
128                        Mac   
169                        Mac   
179                    Windows   
207                        Mac   
228                    Windows   
243                    Windows   
264                        Mac   
293                    Windows   
303                        Mac   
305                    Windows   
351                    Windows   
359                        Mac   
399                        Mac   
424                        Mac   
439                    Windows   
441                    Windows   
462                    Windows   
466                    Windows   
483                        Mac   
492                        Mac   
503                        Mac   
508                        Mac   
509                    Windows   
524                        Mac   
535                        Mac   
537                    Windows   
553                    Windows   
567                    Windows   
605                        Mac   

    Do you like doing your homework when it's released or in the end?  \
31                                        Near the end                  
78                     

### 🔬 Test Case Checkpoint 🔬

In [98]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(len(df_very_happy) == 54 ), "Double check your conditional used to create df_very_happy."
assert(len(df_happy) == 519 ), "Double check your conditional used to create df_happy."
assert(len(df_unhappy) == 36 ), "Double check your conditional used to create df_unhappy."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 2.2: Average Phone Battery Percentage by Group

Now find the **average phone battery percentage** of each group (`df_very_happy`, `df_happy`, and `df_unhappy`):

- The `df['column name'].mean()` function returns the mean of all values in the specified column of `df`.
- The `phone_battery_question` question has been provided to you as a **string** for ease of DataFrame column access.

In [101]:
phone_battery_question = 'What is your phone battery percentage currently at?'

very_happy_people_avg_phone_battery = df_very_happy[phone_battery_question].mean()
very_happy_people_avg_phone_battery

56.96296296296296

In [102]:
happy_people_avg_phone_battery = df_happy[phone_battery_question].mean()
happy_people_avg_phone_battery

59.335907335907336

In [103]:
unhappy_people_avg_phone_battery = df_unhappy[phone_battery_question].mean()
unhappy_people_avg_phone_battery

56.02777777777778

### 🔬 Test Case Checkpoint 🔬

In [104]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(very_happy_people_avg_phone_battery,  56.96296296296296)), "The average phone battery percentage for very happy people does not appear to have been correctly calculated."
assert(math.isclose(happy_people_avg_phone_battery,  59.335907335907336)), "The average phone battery percentage for happy people does not appear to have been correctly calculated."
assert(math.isclose(unhappy_people_avg_phone_battery,  56.02777777777778)), "The average phone battery percentage for unhappy people does not appear to have been correctly calculated."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Analysis: Happiness vs. Phone Battery

**Q: What is the relationship between happiness and phone battery?  Can you think of a possible *confounding variable* in the observed relationship (or lack thereof) between happiness and phone battery?**  Write at least three complete sentences that includes:
- Your analysis of the relationship between the two variables,
- Your third, possibly confounding, variable that might be impacting the results, **AND**
- Your analysis on why this third variable could be a possible confounder.
- Any additional analysis on other variables or possible confounders.

The results of the variables indicate that happy people tend to have higher battery percentages. A third confounding variable could be the fact that the battery life on someone who is unhappy is very low and bad contributing to the low battery. they might be happy and their battery is low but it does not necessarily mean that they are unhappy. 

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)